### ISSUE TITLE
Automatically tune compatibility threshold for species when training

### ISSUE THREAD

**Issue by qwertyuu** [yaricom/goNEAT#48]
Hello! 
While reading [this cool documentation on the technicalities of NEAT](https://www.cs.ucf.edu/~kstanley/neat.html), I stumbled upon a Q/A section named:

### - Is there a way to automatically set the compatibility threshold for speciation?

![image](https://user-images.githubusercontent.com/3145639/158078611-c591a92d-3b9d-46a0-985e-acfb0c2b9071.png)

In their implementation, there seems to be a way to edit the source code and to recompile the NEAT engine so that a certain amount of species is kept when training, so that the number of species does not "explode uncontrollably" when some kind of mutations happen, and even sometimes approach the number of organism in the population, which is by itself very weird (I am the only organism of my species, and so are you! and you.. and you! Is this what extinction feels like?)

I find myself tweaking my threshold beforehand a lot, so that the number of species does not go too high, but it feels like having a set number of species would be desirable. 

Have you thought about implementing this in goNEAT? Thanks!

**Comment by yaricom** [comment]
Hello @qwertyuu 

Yes, I was thinking about this. However, there are no easy way to tune this automatically because evolution is a stochastic process.

Regarding tweaking algorithm. It can be more beneficial to tweak the seed of the random number generator. This can lead to better results. The speciation threshold and other parameters is like environmental factors in natural evolution. The power of evolutionary process is that it can find workarounds if organisms found itself in hostile environment. However, it can take more time to produce the winner.

Anyway, this is good idea. I'll keep it in my to-do list.

Meanwhile, I'm not able to maintain this library in full speed due to **war in Ukraine**, where I'm living.

**Comment by qwertyuu** [comment]
> Meanwhile, I'm not able to maintain this library in full speed due to **war in Ukraine**, where I'm living.

How is the situation? It is all we talk about here. I have donated before and hope that you will win. I am sorry this happens to you.

**Comment by yaricom** [comment]
Yes, this is a war which russia cowardly started against independent country in Europe, that you can see on all the media. We are fighting with all means that we have to protect Ukraine, our families, our freedom. As you know we are fighting very good and already surprised the world. However, any help that can be provided by the civilized world is very much appreciated. 

Thank for your support!

**Comment by qwertyuu** [comment]
Just so everyone reading here wants to donate too: https://itch.io/b/1316/bundle-for-ukraine

This is a game bundle that many game dev DONATED their games so people could donate to Ukraine. So far 5M$ were raised and still going.

Take care @yaricom , we need your brillant brain on this earth :)

**Comment by MHarris021** [comment]
Link shared across my company slack of 500+ and on my linkedin network of 7k+

**Comment by qwertyuu** [comment]
@yaricom I had a thought about you and the situation in Ukraine. Just so you know, you are still on my mind. I wonder how you've been and how the situation is.

**Comment by yaricom** [comment]
@qwertyuu Thank you. I was all the time and still in Kyiv writing code with rockets flying over my head :)
Almost every night for more than a month we have an attacks by cruise, ballistic, etc. missiles accompanied by drones. Sometimes we have it also at the morning when people commute.
But we are keep doing. Working hard and looking forward for the victory.
